osu.Game/Screens/Play/Player.cs

```diff
@@ -527,33 +531,50 @@ private void updateCompletionState(ValueChangedEvent<bool> completionState)
 
             if (!showResults) return;
 
+            scoreSubmissionTask ??= Task.Run(async () =>
+            {
+                var score = CreateScore();
+
+                try
+                {
+                    await SubmitScore(score);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error(ex, "Score submission failed!");
+                }
+
+                try
+                {
+                    await ImportScore(score);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error(ex, "Score import failed!");
+                }
+
+                return score.ScoreInfo;
+            });
+
             using (BeginDelayedSequence(RESULTS_DISPLAY_DELAY))
-                completionProgressDelegate = Schedule(GotoRanking);
+                scheduleCompletion();
         }
 
-        protected virtual ScoreInfo CreateScore()
+        private void scheduleCompletion() => completionProgressDelegate = Schedule(() =>
         {
-            var score = new ScoreInfo
+            if (!scoreSubmissionTask.IsCompleted)
             {
-                Beatmap = Beatmap.Value.BeatmapInfo,
-                Ruleset = rulesetInfo,
-                Mods = Mods.Value.ToArray(),
-            };
-
-            if (DrawableRuleset.ReplayScore != null)
-                score.User = DrawableRuleset.ReplayScore.ScoreInfo?.User ?? new GuestUser();
-            else
-                score.User = api.LocalUser.Value;
-
-            ScoreProcessor.PopulateScore(score);
+                scheduleCompletion();
```

Comment:
I personally preferred the schedule call where it was in the original code (is a bit hidden here if you're not looking carefully).

Also, just for clarity what's the state where the completion state can go from true to false? Is that actually something we need/want to be covering?

Reply:
The way I interpreted this was that this schedule just repeats every frame until the underlying task completes. I don't think it can feasibly ever go from true to false, it'll just be false until (in the timeshift case) the underlying task completion source's result is set.

Reply:
Specifically wonder about this logic:

https://github.com/ppy/osu/pull/11199/files#diff-2711e888291c0df6e400a5c169b90fabe525f3aa8ffe88734eeea89504abb7f2L511-L516

and what it's there for. But based off there being tests covering this it seems like something we want (if a user exits during the completion procedure?)

![image](https://user-images.githubusercontent.com/191335/102706348-6f230c80-42d4-11eb-9e4f-3a7082d0ba1c.png)

Reply:
Yeah those errors are the only reason why the schedule exists. They're also the reason why the submission/import task has a `??=` attached to it.

I didn't feel like breaking it for the time being.